pxr/base/tf/delegatedCountPtr.h

```diff
@@ -0,0 +1,268 @@
+            _RetainIfValid();
+        }
+        return *this;
+    }
+
+    /// Assigns by copying a convertible pointer from `ptr`.
+    ///
+    /// This is necessary in part to allow a `const` `ValueType` to be
+    /// constructed from from a non-`const` `ValueType`.
+    ///
+    /// Retains only if the pointer is valid.
+    template <typename OtherType>
+    TfDelegatedCountPtr& operator=(
+        const TfDelegatedCountPtr<OtherType>& ptr) noexcept {
+        static_assert(_IsPtrConvertible<OtherType>::value);
+        // Does this need to guard against self assignment?
+        _ReleaseIfValid();
+        _pointer = ptr.get();
+        _RetainIfValid();
+        return *this;
+    }
+
+    /// Assigns by moving the pointer from `ptr`.
+    ///
+    /// `ptr` will be reset to its default state (ie. `nullptr`).
+    TfDelegatedCountPtr& operator=(TfDelegatedCountPtr&& ptr) noexcept {
+        if (std::addressof(ptr) != this) {
+            _ReleaseIfValid();
+            _pointer = ptr._pointer;
+            ptr._pointer = nullptr;
+        }
+        return *this;
+    }
+
+    /// Resets the pointer to its default state (ie. `nullptr`)
+    TfDelegatedCountPtr& operator=(std::nullptr_t) noexcept {
+        reset();
+        return *this;
+    }
+
+    /// Releases if the underlying pointer is valid 
+    ~TfDelegatedCountPtr() noexcept {
+        _ReleaseIfValid();
+    }
+
+    /// Dereference the underlying pointer
+    ReferenceType operator*() const { 
+        return *get();
+    }
+
+    /// Arrow operator dispatch for the underlying pointer
+    RawPtrType operator->() const noexcept {
+        return get();
+    }
+
+    /// Returns true if the underlying pointer is non-null.
+    explicit operator bool() const noexcept { return get(); }
+
+    /// Returns true if the underlying pointers are equivalent.
+    template <typename OtherType>
+    bool operator==(
+        const TfDelegatedCountPtr<OtherType>& other) const noexcept {
+        return get() == other.get();
+    }
+
+    /// Returns false if the underlying pointers are equivalent.
+    template <typename OtherType>
+    bool operator!=(
+        const TfDelegatedCountPtr<OtherType>& other) const noexcept {
+        return get() != other.get();
+    }
+
+    /// Orders based on the underlying pointer.
+    template <typename OtherType>
+    bool operator<(
+        const TfDelegatedCountPtr<OtherType>& other)const noexcept {
+        return get() < other.get();
+    }
+
+    /// Return the underlying pointer.
+    RawPtrType get() const noexcept { return _pointer; }
+
+    /// Reset the pointer to its default state (`nullptr`), releasing if
+    /// necessary.
+    void reset() noexcept {
+        _ReleaseIfValid();
+        _pointer = nullptr;
+    }
+
+    /// Swaps the pointer between two delegated count pointers
+    void swap(TfDelegatedCountPtr& other) noexcept {
+        using std::swap;
+        swap(other._pointer, _pointer);
```

Comment:
Do we want ADL here?  I kind of think that we really want to literally swap the pointers.